digdag-server/src/main/java/io/digdag/server/WorkflowExecutionTimeoutEnforcer.java

```diff
@@ -0,0 +1,160 @@
+
+    private final Duration attemptTTL;
+    private final Duration reapingInterval;
+    private final Duration taskTTL;
+
+    @Inject
+    public WorkflowExecutionTimeoutEnforcer(ServerConfig serverConfig, SessionStoreManager ssm, Config systemConfig)
+    {
+        this.attemptTTL = systemConfig.getOptional("executor.attempt_ttl", DurationParam.class)
+                .transform(DurationParam::getDuration)
+                .or(DEFAULT_ATTEMPT_TTL);
+
+        this.taskTTL = systemConfig.getOptional("executor.task_ttl", DurationParam.class)
+                .transform(DurationParam::getDuration)
+                .or(DEFAULT_TASK_TTL);
+
+        this.reapingInterval = systemConfig.getOptional("executor.ttl_reaping_interval", DurationParam.class)
+                .transform(DurationParam::getDuration)
+                .or(DEFAULT_REAPING_INTERVAL);
+
+        this.ssm = ssm;
+
+        if (serverConfig.getExecutorEnabled()) {
+            this.scheduledExecutorService = Executors.newSingleThreadScheduledExecutor(new ThreadFactoryBuilder()
+                    .setDaemon(true)
+                    .setNameFormat("attempt-timeout-enforcer-%d")
+                    .build());
+        }
+        else {
+            this.scheduledExecutorService = null;
+        }
+    }
+
+    private void run()
+    {
+        try {
+            enforceAttemptTTLs();
+        }
+        catch (Throwable t) {
+            logger.error("Uncaught exception", t);
+        }
+
+        try {
+            enforceTaskTTLs();
+        }
+        catch (Throwable t) {
+            logger.error("Uncaught exception", t);
+        }
+    }
+
+    private void enforceAttemptTTLs()
+    {
+        Instant creationDeadline = ssm.getStoreTime().minus(attemptTTL);
+        int state = 0;
+        long lastId = 0;
+
+        List<StoredSessionAttempt> expiredAttempts = ssm.findAttemptsCreatedBeforeWithState(creationDeadline, state, lastId, 100);
+
+        for (StoredSessionAttempt attempt : expiredAttempts) {
+            AttemptStateFlags stateFlags;
+            try {
+                stateFlags = ssm.getAttemptStateFlags(attempt.getId());
+            }
+            catch (ResourceNotFoundException e) {
+                logger.debug("Session Attempt not found, ignoring: {}", attempt, e);
+                continue;
+            }
+
+            if (stateFlags.isCancelRequested()) {
+                logger.debug("Session Attempt already canceled, ignoring: {}", attempt);
+                continue;
+            }
+
+            logger.info("Session Attempt timed out, canceling: {}", attempt);
+            ssm.requestCancelAttempt(attempt.getId());
+        }
+    }
+
+    private void enforceTaskTTLs()
+    {
+        Instant startDeadline = ssm.getStoreTime().minus(taskTTL);
+        long lastId = 0;
+
+        List<TaskAttemptSummary> expiredTasks = ssm.findTasksStartedBeforeWithStateAndType(TaskType.of(0), TaskStateCode.notDoneStates(), startDeadline, lastId, 100);
```

Comment:
I see. Then, how about excluding `PLANNED` state?: `Arrays.stream(notDoneStates()).filter(state -> state != PLANNED).collect(toList())`
I actually don't know which is better. Making `loop>` operator timed out after `taskTTL` is also a possible option.